### DISCUSSION TITLE
Future of System Tray APIs

### DISCUSSION THREAD

**Discussion by TheJoeFin** [microsoft/WindowsAppSDK#519] [General]
I know putting icons in the System Tray is not the blessed path for modern Windows app development, but I am interested in what Microsoft sees as the future for the System Tray APIs.

`Shell_NotifyIcon` is the only way  to add an app into the System Tray but today this requires app to be full trust and as discussed in #219 even though the app does something very small and trivial it has access to the entire PC which could cause big issues to a user due to malice or mistake. 

Many popular apps uses the System Tray to display an icon. 
- OneDrive, Outlook, OneNote (2016), Teams, Skype
- PowerToys
- Spotify
- Zoom
- Discord 
- VPN apps
- Anti-virus apps
- etc.

Is the System Tray outside the scope for Project Reunion? Without resorting to full trust, will the System Tray API be available to developers?

As a Windows user I like the System Tray and find it useful. Also many of my daily productivity apps use the System Tray. I am worried if there is no acknowledgement that this API is popular and should be modernized, developers will continue to ignore the more secure AppContainer model.

**Comment by riverar** [comment]
Add [EarTrumpet](https://eartrumpet.app/) to the list!

Also, some other points from memory, `Shell_NotifyIcon` ...
* doesn't allow for scalable vectors/text glyphs like the modern notification icons do
    * this means DPI changes result in blurry icons for small period of time while the icon is updated, looks bad
* actions (e.g. NIM_MODIFY) can fail in unexpected undocumented ways and is nearly impossible to build a recovery workflow for
    * especially when using `NIF_GUID`, which requires juggling icon identities, polluting the user's system, etc.
    * Skype has a notification area icon that goes blank after certain amount of time passes; if they can't get it right...
* sends you down the old system tooltip path

**Reply by riverar** [reply]
Update: Interestingly, the notification area appears to be getting the undock treatment (or is at least seeing a refactor) in recent builds. Would be great if Microsoft could confirm this is a move in the desired direction.

**Comment by riverar** [comment]
The unfortunate part here is that I'm not aware of any OS provided extensibility points for _rich_ notification area icons. Those nice icons are baked directly into the Shell last I checked. So I think the best Reunion can offer here is a `Shell_NotifyIcon` wrapper. And then there's the User Interface Privilege Isolation (UIPI) and other related concerns to deal with as well.

Maybe the upcoming new taskbar will surprise us? 🙏

**Reply by TheJoeFin** [reply]
I don't expect Project Reunion to bring any enhanced capability to the `Shell_NotifyIcon` but a wrapper to enable System Tray Icons for apps running within an AppContainer seems within the boundaries of the project. The number of desktop bridge apps which use a System Tray Icon is enough evidence that this API is desired by developers and understood by users.  

That being said, a revamped System Tray with new APIs would be incredible!

**Comment by mveril** [comment]
I suggest you look at the issue I posted one year ago in the WinUI repository microsoft/microsoft-ui-xaml#2020. But indeed I think Project Reunion may be a great home for that.
A great thing will also be to use a WinUI context menu which follows the taskbar theme like the system icons do, and allow icon and flyouts to follow this theme as that of OneDrive. I think system Tray notifications APIs also called "balloon tooltip" should be considered as depreciated in favor of Toast Notifications and adaptive cards APIs. But the other system Tray APIs like icons  in the system tray area which can show flyouts and contextmenus should have a future and be modernized with the fluent design.

**Reply by TheJoeFin** [reply]
That is a good discussion you linked. Although worrisome to see people say "we don't need System Tray Icons" even though many modern apps from major companies including Microsoft use them. 

The distinction I'd draw between that discussion and this one is about packaging and Full Trust not user interface. Currently for apps to put an icon in the System Tray they need to be full trust. It would be nice if apps could use the System Tray from within the AppContainer.

**Reply by mveril** [reply]
You're absolutely right.
The need to have a full trust app to deliver a system tray icon is a real problem. This requires some developers to add a full trust Win32 program to their UWP package or to use the Win32 app model just for this when the UWP app model would be sufficient.
Allowing apps with a system tray icon to run in an AppContainer will be a big step forward in rehabilitating this functionality which for me doesn't have a good replacement. Personally I am perplexed when I hear some developer who say that this functionality may be depreciated with windows 10 even if Microsoft and many others continue to use it.

**Comment by TheDecryptor** [comment]
I'd like to see something that could replace the system tray, it feels like a holdover from the days where you couldn't represent a background application with no active windows (Where now we've got things like live tiles and notifications to allow things like that)

Of the applications I've got running that place a system tray icon, only 1-2 actually serve some functionality. The rest are effectively just shortcuts to open their windows without having an icon kept in the taskbar (Like Steam or the Nvidia settings app).

**Reply by TheJoeFin** [reply]
Replacing the System Tray would be a project for the Windows group. In the scope of Project Reunion I am hoping to see the System Tray opened to apps within the AppContainer.

The System Tray is obviously useful to software makers and to users because it continues to be used by old and new software. There are several different use cases, but a major use is to communicate when utility applications are running.

**Comment by ghost** [comment]
[Coming soon](https://github.com/microsoft/ProjectReunion/tree/main/docs#project-reunion-family-components) section in Readme :  
![Screenshot 2021-03-28 181449](https://user-images.githubusercontent.com/74822870/112752497-33938f00-8ff5-11eb-912d-4837a388d317.png)

[Future direction](https://github.com/microsoft/ProjectReunion/tree/main/docs#future-direction)
![Screenshot 2021-03-28 183403](https://user-images.githubusercontent.com/74822870/112752388-9f292c80-8ff4-11eb-8ca7-761d06980e93.png)

these should tell you something

**Comment by soumyamahunt** [comment]
I think the System Tray has some limitations like not a good UX for touch first devices. I think the path to moving forward is new API that allows apps to declare custom widgets for Action Center (like Android allows for their Quick Settings) as it is much easier to use both in desktop and tablet scenarios. 

Here are some features that I would expect this API to have to provide all the features that current system tray provides and more:

1. Apps can declare multiple widgets that user can choose to show in Action Center by editing it, these widgets could have like On/Off mode or just On mode. Apps use these widgets to make it easier for user to toggle some functionality or launch app etc. PowerToys can have multiple of such widgets for toggling different functionality.

2. Optionally apps can provide widgets with additional options to show flyout with them. Basically these widgets will behave like Wifi option. Apps can show anything in the flyout. [EarTrumpet](https://eartrumpet.app/), OneDrive etc. can leverage this feature.

3. Also, providing option to user to enable/disable icons for the widgets in the taskbar itself so users can directly open flyout associated with that widget from that icon without having to go through Action Center. Apps can show anything in the flyout. [EarTrumpet](https://eartrumpet.app/), OneDrive etc. can leverage this feature. The option to showing icons in taskbar can be shown in Windows Settings page for turning system icons on/off or providing a good drag-drop interface where user drags the widget from Action Center to taskbar to create the icon.

These are the scenarios I can think of where all the use cases for tray icons can be covered with a rich experience for both tablet and dektop users.

**Comment by wmail-ak** [comment]
thank you
